### ISSUE TITLE
Enhance virtual support to align around pages of data rather than array of rows

### ISSUE THREAD

**Issue by davoutuk** [swimlane/ngx-datatable#1956]
I'm submitting a ...

```
[ ] bug report => search github for a similar issue or PR before submitting
[x ] feature request
[ ] support request => Please do not submit support request here, post on Stackoverflow or Gitter
```

Current behavior**

The virtual paging support is good but it requires that the underling component include an array of rows.  If the related data source has a very large number of data rows (e.g. 500,000) then the associated data array can get very large for no reason.   For example, if my virtual table has a 1,000 pages with 20 rows per page,  if I scroll to the last page then the associated array requires that array entries be set into array elements 19980 to 19999.   That seems hugely wasteful

An alternative approach would be to support caching of rows by page.  

Expected behavior

Add new virtual paging property:   'virtualDataMethod' that has two possible values:  'byRow'  and 'byPage'
This new property would drive how the table logic asks for data 
If the value is 'byRow' then the logic will use the existing array of rows approach
If the value is 'byPage' then the logic will expect a page/array of rows to be returned. 

This last new method avoids the need to create a large array of rows.   Instead the custom logic would return a page of rows (e.g. an array of rows) for a given page.   It would leave the sourcing of the page of rows to the custom logic

Reproduction of the problem

n/a

What is the motivation / use case for changing the behavior

Adding a new virtual paging method that returns pages of data allows the custom logic to decide on how best to cache data
It also avoids the reqiurement for arrays where very large data tables are to be displayed

Please tell us about your environment:

Windows 10;  Intellij, 

Table version:
n/a

** Angular version:
10

- **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

- **Language:** [all | TypeScript X.X | ES6/7 | ES5]

**Comment by Hypercubed** [comment]
Agreed.  I updated the demo to fix a bug and make the code more clear.  Would be nice to improve the DX around this.
